app/importers/mahonia_input_record.rb

```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+class MahoniaInputRecord < Darlingtonia::InputRecord
+  ##
+  # @return [Hash<Symbol, Object>]
+  def attributes
```

Comment:
Hm, I think this approach might obfuscate the purpose too much, it kind of hides the lookup and translation in the method_missing overwrite. You don't know what the code is doing from reading the method names. I'm not sure that's a good pattern for application code, maintained by in-house developers, because I think less abstraction has more value there.

Reply:
> You don't know what the code is doing from reading the method names.

This is exactly the goal I have in mind. This encapsulation is what enables broad code reuse.

Reply:
I have a PR in on `darlingtonia` that documents this, and might help clarify intent. https://github.com/curationexperts/darlingtonia/pull/8

Reply:
Another idea might be to return hashes of bepress and hyrax from fields, so that we could do something like this in the base input class
` attrs[field[:hyrax_value]] = public_send(field.key)`

which preserves the intent of using the mapper for mapping and doesn't pull in the InputRecord class, and is still readable.

Reply:
I'm not following how the `field[:hyrax_value]` option is more explicit. Can you say a bit more?

Reply:
Because the fields method would return exactly what the attributes method needs, without involving method_missing. The line in the attributes method is trying to use one term for two purposes, and we could just return the two terms appropriate for each purpose. It could be named and structured something like `attrs[field[mapped_hyrax_term.key]] = public_send(mapped_hyrax_term[:key]) `

Reply:
> the line in the attributes method is trying to use one term for two purposes

I think you may be misinterpreting the idea behind `MetadataMapper`. The fields listed by `#fields` are intended only as fields for attributes on the hyrax objects. Mapper is intended to make sure that `InputRecord` does not need to know anything about the contents of the input metadata, so it can focus on providing values (and the attributes hash) that support construction of the savable object.

Reply:
I'm sorry I didn't get back here earlier, I'm getting ready for a class. That line takes a field name, gets its value from the source file (csv) via public_send on the right, and then uses that same field name to store it in the attributes hash it passes to the create, on the left, isn't that correct? That's exactly the place where it seems we want a map (for Mahonia's purposes), of the source file term on the right and the hyrax term on the left. So it seems justified to overwrite that method in Mahonia, in a MahoniaInputRecord class, without compromising the separation of concerns in Darlingtonia.

Do you want to get into more detail about what you'd rather see, in an email or Slack message? Are you suggesting we should keep a MahoniaInputRecord class and transfer what's in the MahoniaMapper into it, and get rid of the MahoniaMapper? I could check it out when I get back and work with it in the morning.

Reply:
No worries about timing, we have lots of time to sort this out properly. \

> That line takes a field name, gets its value from the source file (csv) via public_send on the right, and then uses that same field name to store it in the attributes hash it passes to the create, on the left, isn't that correct?

The `InputRecord` doesn't read the CSV. Responsibility for that is pushed to the the `Mapper`. The `Mapper`'s job is to accept a metadata object, say what fields it offers, and do mapping between the metadata object and the fields. The `InputRecord` does not know about the CSV, even to the point of knowing more about it than that it is an `Object` that can be passed to a mapper.

To my eyes, the `Mahonia` specific behavior we need is a mapping from a CSV, maybe with assumptions about fields, or transformation of specific values. The `InputRecord` in the base is already written to be agnostic about the details of the metadata input and the mapping, so I'm concerned that our solution involves rewriting its most significant behavior. If we feel we need to do that, I'd want to think about a rewrite of the base class that avoids the need for custom behavior.

> Are you suggesting we should keep a MahoniaInputRecord class and transfer what's in the MahoniaMapper into it, and get rid of the MahoniaMapper?

I'm not suggesting this; however, if our opinion in implementing the mahonia importer is that being explicit is better, I don't see any benefit of the `Mapper` abstraction. If we are tending toward a less abstract implementation, I'd want to just inline the mapper behavior so we don't have a costly, obfuscating abstraction that doesn't give us the code reusability we are looking for.

Reply:
I think there is something not quite right about how these classes are expected to interact, InputRecord and the Mapper. I think the MahoniaMapper must be not doing its fundamental job. I think it might need to translate the metadata row it gets into a hash with the Hyrax field names, store the original hash but make an attributes hash. The fields method needs to return attribute keys and this in method_missing `return Array(metadata[method_name.to_s]) if fields.include?(method_name)` needs to access the attribute hash. Maybe I need to overwrite the Mapper `metadata=(meta)` method to do the mapping, use the `map_bepress_terms` in it, something like that.

Reply:
> I think the MahoniaMapper must be not doing its fundamental job.

Yes, it's not providing #fields, and not offering methods for the fields. The issues may be worse than that, but I would like to resolve this issue to begin with.